### PR TITLE
add Z.divide support for Z.to_euclidean_division_equations

### LINKED ISSUE #17918: `Z.to_euclidean_division_equations` should support `Z.divide`
Mostly as a reminder to myself to implement this (though someone else is welcome to take it), it should be pretty easy to get `Z.to_euclidean_division_equations` to support `Z.divide`, which is defined as `Z.divide = fun x y : Z => exists z : Z, y = (z * x)%Z`, just by adding, e.g., `Ltac divide_to_equations_step` that does something like `match goal with | [ H : Z.divide _ _ |- _ ] => destruct H end`, then adding the corresponding helper tactics, in `theories/omega/PreOmega.v`

### PR DESCRIPTION
**Pull request by doctor-kaliy**
Fixes #17918 according to the description provided in the Issue.

~~Depends on #17934.~~

### CODE REVIEW

**Review by JasonGross — CHANGES_REQUESTED** [review]
Thank you!  Please add a small test to the test suite at the bottom of
https://github.com/coq/coq/blob/d95cb47c4f338b8ee18dcbcb5f18ba5622cfe028/test-suite/success/Nia.v#L919 (something like `Lemma test_divide : forall x y, Z.divide x y -> x mod 0 = 0. Proof. intos; nia. Qed.`, maybe?), add a changelog entry with `dev/tools/make-changelog.sh`, and I'll be happy to approve and merge

**Comment by doctor-kaliy** [comment]
> Thank you! Please add a small test to the test suite at the bottom of
> 
> https://github.com/coq/coq/blob/d95cb47c4f338b8ee18dcbcb5f18ba5622cfe028/test-suite/success/Nia.v#L919
> (something like `Lemma test_divide : forall x y, Z.divide x y -> x mod 0 = 0. Proof. intos; nia. Qed.`, maybe?), add a changelog entry with `dev/tools/make-changelog.sh`, and I'll be happy to approve and merge

Thank you for your feedback.

### Here is what I found out trying to add some tests

This could have been a good test ``` Lemma Z_divide_test (x y : Z) (H : (y | x)) : x <> 0 \/ y <> 0 -> x mod y = 0. ```
I tried to prove it by ``` intros; nia. ``` and it didn't work. ``` Tactic failure:  Cannot find witness. ```
But this lemma can be proved by ``` intros. Z.to_euclidean_division_equations; assert (x0 = q); nia. ```(or by any other "by hand proof") Here we prove that ```x0``` we get from destructing ```(y | x)```, and ```q``` from ```Z.div_mod_to_equations'``` are equal.

Though, nia is able to handle cases like:
```
Lemma Z_divide_test_5 (z: Z) (H : (5 | z)) : z mod 5 = 0. Proof. intros. nia. Qed.
Lemma Z_divide_test_neg_5 (z: Z) (H : ((-5) | z)) : z mod (-5) = 0. Proof. intros. nia. Qed.
Lemma Z_divide_test_5_neg_5 (z: Z) (H : (5 | z)) : z mod (-5) = 0. Proof. intros. nia. Qed.
```
but these don't look meaningful enough :( .

Now I'm a bit confused about the test to add. Tests with constant values are too simple and test that checks divisibility can't be proved automatically.

### One more thing worth flagging
 ```Lemma Z_divide_test' (x y : Z) : Z.rem x y = 0 -> x mod y = 0. Proof. intros. nia. Qed.``` also doesn't work. It can be done with ```Z.to_euclidean_division_equations; assert (q0 = q); nia.```, for instance.

**Comment by JasonGross** [comment]
That's a good point.  I've created https://github.com/coq/coq/pull/17934 to assert such "equal quotient" equations when they are obvious.  If you add
```coq
(* Make the direction of [Z.divide] line up with the rest of the Euclidean equation facts *)
Local Lemma divide_alt x y : Z.divide x y -> exists z, y = x * z.
Proof. intros [z H]; exists z; subst; apply Z.mul_comm. Qed.
```
and handle `Z.divide` with
```coq
match goal with
| [ H : Z.divide _ _ |- _ ] => try apply divide_alt in H; destruct H
end.
```
then the improvement from #17934 should be sufficient to handle it.  I guess without #17934 and without a fix for #17935 (I was considering that `(x | y) -> (y | x) -> x = y \/ x = -y` might be good, but `nia` can't even prove `x * y = 1 -> x = 1 \/ x = -1`...) this PR doesn't actually add more power, so we should wait for #17934.  (Or you can rebase on #17934, but #17934 breaks the test-suite and I've run out of time today to fix it.)

Also, would you add ```:g:`Z.divide` ``` to the list at https://github.com/coq/coq/blob/e78e90d219e66ce00bf9e78e228ac51b5ab17020/doc/sphinx/addendum/micromega.rst?plain=1#L371

**Comment by JasonGross** [comment]
This is blocked on https://github.com/coq/coq/pull/17934.  Once https://github.com/coq/coq/pull/17934 is merged all this needs is to make sure test-suite examples work

**Comment by JasonGross** [comment]
Now that https://github.com/coq/coq/pull/17934 has finally been merged, if you want to rebase this @doctor-kaliy it should be possible to add some actual tests.

File: doc/changelog/04-tactics/17927-divide_to_euclidean_division_equations.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Added:**
+  When using :g:`Z.to_euclidean_division_equations` can now pose
+  equations of the form ``x = y * q`` using :g:`Z.divide`.
+  (`#17927 <https://github.com/coq/coq/pull/17927>`_,
+  by doctor-kaliy).
```
Comment (jfehrle) [inline doc/changelog/04-tactics/17927-divide_to_euclidean_division_equations.rst:5]:
```suggestion
  When using :g:`Z.to_euclidean_division_equations`, you can now pose
  equations of the form ``x = y * q`` using :g:`Z.divide`.
  (`#17927 <https://github.com/coq/coq/pull/17927>`_,
    by Evgenii Kosogorov).
```

Nice to use real names.

**Review by jfehrle — APPROVED** [review]
wording suggestions

**Review by JasonGross — APPROVED** [review]

**Comment by JasonGross** [comment]
@coqbot run full ci

Once you respond to @jfehrle 's review, assuming CI passes, I'll plan to merge by Monday

**Comment by JasonGross** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
machine was missing python-is-python3
@coqbot bench

**Comment by JasonGross** [comment]
@coqbot run full ci

**Comment by JasonGross** [comment]
@coqbot run full ci

File: doc/changelog/04-tactics/17927-divide_to_euclidean_division_equations.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- **Added:**
+  When using :g:`Z.to_euclidean_division_equations`, you can now pose
+  equations of the form ``x = y * q`` using :g:`Z.divide`.
```
Comment (JasonGross) [inline doc/changelog/04-tactics/17927-divide_to_euclidean_division_equations.rst:3]:
```suggestion
  equations of the form ``x = y * q`` using :g:`Z.divide`
```

**Review by JasonGross — CHANGES_REQUESTED** [review]

File: doc/changelog/04-tactics/17927-divide_to_euclidean_division_equations.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+  (`#17927 <https://github.com/coq/coq/pull/17927>`_,
+    by Evgenii Kosogorov).
```
[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/changelog/04-tactics/17927-divide_to_euclidean_division_equations.rst:5]: I believe so.  I suppose `unreleased.rst` is created by some script from this file and others.

**Review by JasonGross — APPROVED** [review]

**Comment by JasonGross** [comment]
quickchick is also failing in the same way on master, so let's merge.  Thanks @doctor-kaliy !

@coqbot merge now

**Comment by JasonGross** [comment]
@coqbot merge now

**Comment by JasonGross** [comment]
Oh, oops, I guess this should not be targeted at 8.19, but at 8.20

**Pull request merged**